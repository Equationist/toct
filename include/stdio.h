/* stdio.h - Standard Input/Output Library for CompilerKit C Compiler */

#ifndef _STDIO_H
#define _STDIO_H

/* Standard I/O types */
typedef struct {
    int _fd;
    int _flags;
    char* _buffer;
} FILE;

/* Standard streams */
extern FILE* stdin;
extern FILE* stdout; 
extern FILE* stderr;

/* Standard I/O functions */
int printf(const char* format, ...);
int fprintf(FILE* stream, const char* format, ...);
int sprintf(char* str, const char* format, ...);
int snprintf(char* str, int size, const char* format, ...);

int scanf(const char* format, ...);
int fscanf(FILE* stream, const char* format, ...);
int sscanf(const char* str, const char* format, ...);

int putchar(int c);
int puts(const char* s);
int fputs(const char* s, FILE* stream);
int fputc(int c, FILE* stream);

int getchar(void);
char* gets(char* s);
char* fgets(char* s, int size, FILE* stream);
int fgetc(FILE* stream);
int ungetc(int c, FILE* stream);

FILE* fopen(const char* filename, const char* mode);
FILE* freopen(const char* filename, const char* mode, FILE* stream);
int fclose(FILE* stream);
int fflush(FILE* stream);

long ftell(FILE* stream);
int fseek(FILE* stream, long offset, int whence);
void rewind(FILE* stream);
int feof(FILE* stream);
int ferror(FILE* stream);
void clearerr(FILE* stream);

/* File position constants */
#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

/* End of file constant */
#define EOF (-1)

/* Buffer size */
#define BUFSIZ 8192

/* Null pointer */
#ifndef NULL
#define NULL ((void*)0)
#endif

#endif /* _STDIO_H */