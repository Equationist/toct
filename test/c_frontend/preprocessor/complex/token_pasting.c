/* Test token pasting operator (##) */
#define CONCAT(a, b) a##b
#define MAKE_FUNC(name) int func_##name() { return name##_value; }
#define DECLARE_VAR(type, name) type var_##name
#define PRINT_FIELD(struct_name, field) printf(#struct_name "." #field " = %d\n", struct_name.field_##field)

#include <stdio.h>

/* Use token pasting to create variables and functions */
int one_value = 10;
int two_value = 20;
int test_value = 99;

MAKE_FUNC(one)
MAKE_FUNC(two)
MAKE_FUNC(test)

struct MyStruct {
    int field_x;
    int field_y;
    int field_count;
};

int main() {
    DECLARE_VAR(int, counter) = 5;
    DECLARE_VAR(float, ratio) = 3.14f;
    
    printf("Token pasting tests:\n");
    printf("CONCAT(hel, lo) -> hello\n");
    printf("var_counter = %d\n", var_counter);
    printf("var_ratio = %.2f\n", var_ratio);
    
    printf("\nFunction calls via token pasting:\n");
    printf("func_one() = %d\n", func_one());
    printf("func_two() = %d\n", func_two());
    printf("func_test() = %d\n", func_test());
    
    struct MyStruct s = {100, 200, 42};
    printf("\nStruct field access:\n");
    PRINT_FIELD(s, x);
    PRINT_FIELD(s, y);
    PRINT_FIELD(s, count);
    
    return 0;
}