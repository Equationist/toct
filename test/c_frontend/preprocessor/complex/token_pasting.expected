Token pasting tests:
CONCAT(hel, lo) -> hello
var_counter = 5
var_ratio = 3.14

Function calls via token pasting:
func_one() = 10
func_two() = 20
func_test() = 99

Struct field access:
s.x = 100
s.y = 200
s.count = 42